atspi-connection/src/p2p.rs

```diff
@@ -0,0 +1,768 @@
+	/// # Errors
+	/// - `DBusProxy` cannot be created.
+	/// - The socket address cannot be parsed.
+	///
+	pub(crate) async fn try_new<B, S>(
+		bus_name: B,
+		socket: S,
+		conn: &zbus::Connection,
+	) -> Result<Self, AtspiError>
+	where
+		B: Into<OwnedBusName>,
+		S: TryInto<Address>,
+	{
+		let dbus_proxy = DBusProxy::new(conn).await?;
+		let owned_bus_name: OwnedBusName = bus_name.into();
+
+		// Because D-Bus does not let us query whether a unique name is the owner of a well-known name,
+		// we need to query all well-known names and their owners, and then check if the unique name is one of them.
+
+		// Get all well-known names from D-Bus
+		let well_known_names: Vec<OwnedWellKnownName> = dbus_proxy
+			.list_names()
+			.await?
+			.into_iter()
+			.filter_map(|name| {
+				if let BusName::WellKnown(well_nown_name) = name.clone().inner() {
+					Some(OwnedWellKnownName::from(well_nown_name.clone()))
+				} else {
+					None
+				}
+			})
+			.collect();
+
+		// We are creating a mapping of unique names to well-known names.
+		let mut unique_to_well_known: Vec<(OwnedUniqueName, OwnedWellKnownName)> = Vec::new();
+
+		// For each well-known name, we get the unique name of the owner.
+		// Note: not all well-known names on the bus will have an accessible owner.
+		// For instance, the `org.freedesktop.DBus` well-known name does not have an accessible connection.
+		for well_known_name in &well_known_names {
+			let bus_name = BusName::from(well_known_name.clone());
+			if let Ok(unique_name) = dbus_proxy.get_name_owner(bus_name).await {
+				unique_to_well_known.push((unique_name, well_known_name.clone()));
+			}
+		}
+
+		// Now we have a mapping of unique names to well-known names.
+		//
+		// A `Peer` instance requires a unique name which may or may not have a well-known name.
+		// We can build a `Peer` instance from either a unique name or a well-known name.
+		// If the argument name is a unique name, we look up whether this peer also owns a well-known name in our mapping.
+		// If the argument name is a well-known name, we _must_ get its unique owner too to create a `Peer`.
+		let (unique_name, well_known_name) = match owned_bus_name.inner() {
+			BusName::Unique(name) => {
+				// The argument name is the mandatory `UniqueName`, we do want check whether this peer also owns a well-known name.
+				let owned_well_known_name = unique_to_well_known.iter().find_map(|(u, w)| {
+					if u == name {
+						Some(w.clone())
+					} else {
+						None
+					}
+				});
+				let owned_unique_name = OwnedUniqueName::from(name.clone());
+				(owned_unique_name, owned_well_known_name)
+			}
+			BusName::WellKnown(well_known_name) => {
+				let bus_name = BusName::from(well_known_name.clone());
+				let owned_unique_name = dbus_proxy.get_name_owner(bus_name).await?;
+				let owned_well_known_name = OwnedWellKnownName::from(well_known_name.clone());
+				(owned_unique_name, Some(owned_well_known_name))
+			}
+		};
+
+		let socket_address = socket
```

Comment:
This should be done higher up; since it is not dependent on the other portions of the function finishing correctly.

If possible, this error should be able to bubble up before any bus activity. It makes errors potentially more consistent.

Reply:
Generally I am against doing work before earlier bail-outs could happen, but here I agree. Moved up.